### PR TITLE
docs: document spark

### PR DESCRIPTION
**Pull request by sauyon**

### CODE REVIEW

File: docs/source/integrations/spark.rst [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+version 1.0.13 and above and Spark version 3.3.0 and above.
+
+The following tutorial assumes you have a BentoML service ready and know the basics of BentoML. If
+you'd like to learn more about BentoML, see the :ref:`BentoML tutorial <tutorial>`. Also, ensure that
+you have at least BentoML 1.0.13 and Spark version 3.3.0.
+
+For this example, we'll be using the quickstart bento from the aforementioned tutorial, but the
+commands should work for bentos with IO descriptors which support batch inference (at the time of
+writing, those are ``bentoml.io.NumpyNdarray``, ``bentoml.io.PandasDataFrame`` and
+``bentoml.io.PandasSeries``) with the following caveat:
```
Comment (aarnphm) [inline docs/source/integrations/spark.rst:15]:
```suggestion

The following tutorial will use the quickstart bento from :ref:`aforementioned tutorial <tutorial>`.

.. note::

	All of the following commands/APIs should work for bentos with IO Descriptor <link_to_io_descriptor_rst> that supports 

For this example, we'll be using the quickstart bento from the aforementioned tutorial, but the
commands should work for bentos with IO descriptors which support batch inference (at the time of writing, those are ``bentoml.io.NumpyNdarray``<link_to_io_descriptor_numpy>, ``bentoml.io.PandasDataFrame``<link_to_io_descriptor_pandas> and
``bentoml.io.PandasSeries``<link_to_io_descriptor_series>) with the following caveat:
```

Comment (aarnphm) [reply docs/source/integrations/spark.rst:15]: what are the caveats?

File: docs/source/integrations/spark.rst [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+IMPORTANT: your Bento API must be capable of accepting multiple inputs. For example,
```
Comment (aarnphm) [inline docs/source/integrations/spark.rst:17]:
```suggestion
:bdg-warning:`IMPORTANT:` your Bento API must be capable of accepting multiple inputs. For example,
```

File: docs/source/integrations/spark.rst [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        spark = SparkSession.builder.getOrCreate()
+
+#. Load the input data into a PySpark DataFrame. If you are using multipart input, or your dataframe
+   requires column names, you must also provide a schema for your DataFrame as you load it. You can
+   do this using the ``spark.read.csv()`` method, which takes a file path as input and returns a
+   DataFrame containing the data from the file.
+
+    .. code-block:: python
+        from pyspark.sql.types import StructType, StructField, FloatType, StringType
+        schema = SparkStruct(
```
Comment (aarnphm) [inline docs/source/integrations/spark.rst:38]:
```suggestion
    .. code-block:: python
    
        from pyspark.sql.types import StructType, StructField, FloatType, StringType
        schema = SparkStruct(
```

File: docs/source/integrations/spark.rst [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    * First, the bento is distributed to the cluster. Note that if the bento has already been
+      distributed, i.e. you have already run a computation with that bento, this step is skipped.
+
+    * Next, a process function is created, which starts a BentoML server on each of the Spark
+      workers, then uses a client to process all the data. This is done so that the workers take
+      advantage of the batch processing features of the BentoML server. PySpark pickles this process
+      function and dispatches it, along with the relevant data, to the workers.
+
+    * Finally, the function is evaluated on the given dataframe. Once all methods that the user
+      defined in the script have been executed, the data is returned to the master node.
```
Comment (aarnphm) [inline docs/source/integrations/spark.rst:77]: Makes this either a note, dropdown or highlight

File: docs/source/reference/batch.rst [hunk 5]
```diff
@@ -0,0 +1,8 @@
+======
+Batch Inference
+======
```
Comment (aarnphm) [inline docs/source/reference/batch.rst:3]:
```suggestion
===============
Batch Inference
===============
```

File: docs/source/reference/batch.rst [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+
+Batch Inference Backends
+------------------------
+
+.. autofunction bentoml.batch.run_in_spark
```
Comment (aarnphm) [inline docs/source/reference/batch.rst:8]:
```suggestion
.. autofunction:: bentoml.batch.run_in_spark
```

File: src/bentoml/_internal/batch/spark.py [hunk 7]
```diff
@@ -99,6 +99,54 @@ def run_in_spark(
...
+        The name of the inference API to run. If not provided, there must be only one API contained
+        in the bento; that API will be run.
+    output_schema:
+        The Spark schema of the output DataFrame. If not provided, BentoML will attempt to infer the
+        schema from the output descriptor of the inference API.
+
+    Returns:
+        The result of the inference API run on the input ``df``.
+
+    Examples
```
Comment (aarnphm) [inline src/bentoml/_internal/batch/spark.py:125]: Lets use Google style instead of Numpy

File: src/bentoml/_internal/batch/spark.py [hunk 8]
```diff
@@ -99,6 +99,54 @@ def run_in_spark(
...
+        >>> spark = SparkSession.builder.getOrCreate()
+        >>> schema = StructType([
+        ...     StructField("name", StringType(), True),
+        ...     StructField("age", StringType(), True),
+        ... ])
+        >>> df = spark.createDataFrame([("John", 30), ("Mike", 25), ("Sally", 40)], schema)
+        >>>
+        >>> bento = bentoml.get("my_service:latest")
+        >>> results = bentoml.batch.run_in_spark(bento, df, spark)
+        >>> results.show()
```
Comment (aarnphm) [inline src/bentoml/_internal/batch/spark.py:143]: same here, we can also just remove the >>> or our copy button won't work?

File: src/bentoml/_internal/batch/spark.py [hunk 9]
```diff
@@ -99,6 +99,54 @@ def run_in_spark(
...
+        +-----+---+
+        | name|age|
+        +-----+---+
+        |John |30 |
+        +-----+---+
```
Comment (aarnphm) [inline src/bentoml/_internal/batch/spark.py:148]: probably needs ... format here as well

File: src/bentoml/_internal/batch/spark.py (hunk 8 above)
Comment (ssheng) [reply src/bentoml/_internal/batch/spark.py:143]: @sauyon I will merge it now please work with Aaron to address later.

**Review by ssheng — APPROVED** [review]

**Comment by ssheng** [comment]
@sauyon I will merge it now please work with Aaron to address the unresolved comments later.

File: docs/source/integrations/spark.rst [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+    .. code-block:: python
+        from pyspark.sql.types import StructType, StructField, FloatType, StringType
+        schema = SparkStruct(
+            StructField(name=”sepal_length”, FloatType(), False),
+            StructField(name=”sepal_width”, FloatType(), False),
+            StructField(name=”petal_length”, FloatType(), False),
+            StructField(name=”petal_width”, FloatType(), False),
+        )
+        df = spark.read.csv("https://docs.bentoml.org/en/latest/integrations/spark/input.csv")
```
Comment (ssheng) [inline docs/source/integrations/spark.rst:55]: Are we missing this file in this commit?

File: docs/source/reference/batch.rst [hunk 11]
```diff
@@ -0,0 +1,8 @@
+===============
+Batch Inference
+===============
+
+Batch Inference Backends
+------------------------
+
+.. autofunction:: bentoml.batch.run_in_spark
```
Comment (ssheng) [inline docs/source/reference/batch.rst:8]: This is not show up correctly. @sauyon

File: docs/source/integrations/spark.rst (hunk 10 above)
Comment (aarnphm) [reply docs/source/integrations/spark.rst:55]: https://github.com/bentoml/BentoML/pull/3459 address this

**Pull request merged**